File: frontend/lib/src/models/history.dart
enum Action { DOOR_OPENING, DOOR_CLOSING, LOCK_OPENING, LOCK_CLOSING, QR_CODE_SCAN, NFC_SCAN }

class History {
  final int id;
  final int waterClosetId;
  final String datetime;
  final Action action;

  History({required this.id, required this.waterClosetId, required this.datetime, required this.action});

  factory History.fromJson(Map<String, dynamic> json) {
    return History(
      id: json['id'],
      waterClosetId: json['water_closet_id'],
      datetime: json['datetime'],
      action: Action.values.firstWhere((e) => e.toString() == 'Action.' + json['action']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'water_closet_id': waterClosetId,
      'datetime': datetime,
      'action': action.toString().split('.').last,
    };
  }
}

File: backend/src/models/history.rs
use crate::models::water_closet::WaterCloset;
use chrono::NaiveDateTime;
use diesel::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(diesel_derive_enum::DbEnum, Debug, PartialEq, Serialize, Deserialize, Clone)]
#[ExistingTypePath = "crate::schema::sql_types::Actions"]
pub enum Actions {
    #[db_rename = "DOOR_OPENING"]
    DoorOpening,
    #[db_rename = "DOOR_CLOSING"]
    DoorClosing,
    #[db_rename = "LOCK_OPENING"]
    LockOpening,
    #[db_rename = "LOCK_CLOSING"]
    LockClosing,
    #[db_rename = "QR_CODE_SCAN"]
    QRCodeScan,
    #[db_rename = "NFC_SCAN"]
    NFCScan,
}

#[derive(
    Queryable, Selectable, Identifiable, Associations, PartialEq, Debug, Serialize, Deserialize,
)]
#[diesel(belongs_to(WaterCloset))]
#[diesel(table_name = crate::schema::histories)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct History {
    pub id: i32,
    pub water_closet_id: i32,
    pub datetime: NaiveDateTime,
    pub action: Actions,
}

#[derive(Insertable, Serialize, Deserialize)]
#[diesel(belongs_to(WaterCloset))]
#[diesel(table_name = crate::schema::histories)]
pub struct NewHistory {
    pub water_closet_id: i32,
    pub datetime: NaiveDateTime,
    pub action: Actions,
}
